[DOM] For range.extractContents(), abort early if there's a doctype in range

David-Li-Jy opened the pull request:
#### c20d51c646be933823b7b50b3f9c1a4a36950fa7
<pre>
[DOM] For range.extractContents(), abort early if there&apos;s a doctype in range
<a href="https://bugs.webkit.org/show_bug.cgi?id=252805">https://bugs.webkit.org/show_bug.cgi?id=252805</a>
rdar://103178567

Reviewed by Ryosuke Niwa.

Check if doctype is contained in range before start extracting contents,
if contained, return HierarchyRequestError.

* LayoutTests/fast/dom/Range/range-commonroot-notfound-expected.txt: Added.
* LayoutTests/fast/dom/Range/range-commonroot-notfound.html: Added.
* Source/WebCore/dom/Range.cpp:
(WebCore::Range::processContents):

Canonical link: <a href="https://commits.webkit.org/261342@main">https://commits.webkit.org/261342@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6eca9dc2528e3fb2240a13cf033819d289c28d79

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111292 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20435 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43837 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/2722 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120100 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115240 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21800 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11534 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/2400 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117057 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16183 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99369 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/103929 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98141 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/31021 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/44780 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12934 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/32361 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/86620 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13452 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9352 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18891 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51945 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15407 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4304 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on LayoutTests/fast/dom/Range/range-commonroot-notfound.html:
```diff
@@ -0,0 +1,54 @@
+<html>
+<head>
+<script>
+function gc() {
```
Please include ../../../resources/gc.js instead.

David-Li-Jy replied:
Updated.

rniwa commented on Source/WebCore/dom/Range.cpp:
```diff
@@ -310,7 +310,8 @@ ExceptionOr<RefPtr<DocumentFragment>> Range::processContents(ActionType action)
     RefPtr<Node> commonRoot = commonAncestorContainer();
-    ASSERT(commonRoot);
+    if (!commonRoot)
+        return Exception { NotFoundError };
```
Hm... this doesn't seem right. Chrome & Firefox wouldn't throw an exception in this case.
Instead, setStartAfter/setEndBefore are not setting start/end to x1/x2.

David-Li-Jy replied:
Working on it.

rniwa commented on Source/WebCore/dom/Range.cpp:
```diff
@@ -311,6 +311,14 @@ ExceptionOr<RefPtr<DocumentFragment>> Range::processContents(ActionType action)
     RefPtr<Node> commonRoot = commonAncestorContainer();
     ASSERT(commonRoot);
+    
+    // For extractContents(), abort early if there's a doctype in range.
```
This comment just repeats what the code says. Please remove.

David-Li-Jy replied:
Done, thanks!

rniwa commented on LayoutTests/fast/dom/Range/range-commonroot-notfound.html:
```diff
@@ -0,0 +1,49 @@
+<!DOCTYPE html>
```
What if you removed this DOCTYPE. Wouldn't the test case still crash?

David-Li-Jy replied:
No, it doesn't crash.

rniwa commented on Source/WebCore/dom/Range.cpp:
```diff
@@ -311,6 +311,13 @@ ExceptionOr<RefPtr<DocumentFragment>> Range::processContents(ActionType action)
+    
+    if (action == Extract) {
+        Document& commonRootDocument = commonRoot->document();
+        RefPtr<Node> doctype = reinterpret_cast<Node*>(commonRootDocument.doctype());
```
Why do we need to do `reinterpret_cast` here?? That seems totally unnecessary since `DocumentType*` is `Node*`.
Maybe you need to include `DocumentType.h` in `Range.cpp`?

David-Li-Jy replied:
Ahh, thanks! I was also confused before on failing to covert between. Updated.

rniwa commented on Source/WebCore/dom/Range.cpp:
```diff
@@ -311,6 +312,13 @@ ExceptionOr<RefPtr<DocumentFragment>> Range::processContents(ActionType action)
     ASSERT(commonRoot);
+    
+    if (action == Extract) {
+        Document& commonRootDocument = commonRoot->document();
```
We can use auto& instead.

rniwa commented on Source/WebCore/dom/Range.cpp:
```diff
@@ -311,6 +312,13 @@ ExceptionOr<RefPtr<DocumentFragment>> Range::processContents(ActionType action)
+    
+    if (action == Extract) {
+        Document& commonRootDocument = commonRoot->document();
+        RefPtr<Node> doctype = commonRootDocument.doctype();
```
This can just be `RefPtr` instead of `RefPtr<Node>`.

David-Li-Jy committed:
[DOM] For range.extractContents(), abort early if there's a doctype in range
https://bugs.webkit.org/show_bug.cgi?id=252805
rdar://103178567

Reviewed by Ryosuke Niwa.

Check if doctype is contained in range before start extracting contents,
if contained, return HierarchyRequestError.

* LayoutTests/fast/dom/Range/range-commonroot-notfound-expected.txt: Added.
* LayoutTests/fast/dom/Range/range-commonroot-notfound.html: Added.
* Source/WebCore/dom/Range.cpp:
(WebCore::Range::processContents):

Canonical link: https://commits.webkit.org/261342@main
